### PR TITLE
feat: Add Contact Us API endpoint #101

### PR DESCRIPTION
**Pull request by deolla**
## What does the PR do?
This PR adds a new API endpoint to handle user inquiries from the "Contact Us" form. It includes server-side logic for processing inquiries, validation and sanitization of input data, and sending emails to a specified address.

## How should this be manually tested?
- Send a POST request to `/api/v1/contact-us` with the following JSON payload:
  ```json
  {
    "name": "String",
    "email": "String",
    "message": "String"
  }
## Checklist of what you did
- [x] Developed server-side logic to handle "Contact Us" form submissions.
- [x] Implemented validation and sanitization of incoming user data.
- [x] Configured email sending to a specified address using an email service.
- [x] Wrote unit tests to verify that the "Contact Us" endpoint validates input correctly and sends emails
- [x] Tested various scenarios for submitting the "Contact Us" form to cover edge cases.
 
## Reference Issue
[FEAT] Contact Us API Endpoint #101

### CODE REVIEW

File: src/modules/contactUs/contact-us.controller.ts [hunk 1]
```diff
@@ -0,0 +1,25 @@
+import { Body, Controller, HttpException, HttpStatus, Post } from '@nestjs/common';
+import { ContactService } from './contact-us.service';
+import { ContactUsDto } from './dto/contact-us.dto';
+
+@Controller('api/v1/contact-us')
+export class ContactController {
+  constructor(private readonly contactService: ContactService) {}
```
Comment (Homoakin619) [inline src/modules/contactUs/contact-us.controller.ts:7]: I believe this route is a public route. If yes add the decorator @skipAuth() to this endpoint

File: src/modules/contactUs/contact-us.controller.ts
Comment (Homoakin619) [inline src/modules/contactUs/contact-us.controller.ts:1]: move your try catch block to your service

File: src/modules/contactUs/contact-us.service.ts [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+    try {
+      await this.transporter.sendMail(mailOptions);
+    } catch (error) {
+      throw new InternalServerErrorException('Failed to send email');
+    }
+  }
+}
+
+// import { Injectable, InternalServerErrorException } from '@nestjs/common';
```
Comment (Homoakin619) [inline src/modules/contactUs/contact-us.service.ts:41]: Remove the commented code

File: src/modules/contactUs/entities/contact-us.entity.ts
Comment (Homoakin619) [inline src/modules/contactUs/entities/contact-us.entity.ts:1]: This Contact Us Entity should extend the AbstractBaseEntity

File: src/modules/contactUs/entities/contact-us.entity.ts [hunk 3]
```diff
@@ -0,0 +1,20 @@
+import { Column, Entity } from 'typeorm';
+//import { AbstractBaseEntity } from '../../../entities/base.entity';
+
+@Entity()
+export class ContactUs {
+  @Column({ primary: true, generated: 'uuid' })
```
Comment (Homoakin619) [inline src/modules/contactUs/entities/contact-us.entity.ts:6]: remove this

File: src/modules/contactUs/entities/contact-us.entity.ts [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+  @Column({ nullable: false })
+  name: string;
+
+  @Column({ unique: true, nullable: false })
+  email: string;
+
+  @Column({ nullable: false })
+  message: string;
+
+  @Column({ nullable: false })
```
Comment (Homoakin619) [inline src/modules/contactUs/entities/contact-us.entity.ts:18]: remove this

**Comment by deolla** [comment]
Changes have been made

File: src/modules/contactUs/contact-us.service.ts [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    } catch (error) {
+      throw new HttpException(
+        {
+          message: 'Failed to send email',
+          errors: [error.message],
+          status: HttpStatus.INTERNAL_SERVER_ERROR,
+        },
+        HttpStatus.INTERNAL_SERVER_ERROR
+      );
+      //throw new InternalServerErrorException('Failed to send email');
```
Comment (Homoakin619) [inline src/modules/contactUs/contact-us.service.ts:44]: remove commented code

**Comment by deolla** [comment]
done

**Review by Homoakin619 — APPROVED** [review]

**Review by Heba-WebDev — APPROVED** [review]
Good job!

File: src/modules/contactUs/contact-us.service.ts [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+        pass: process.env.EMAIL_PASS,
+      },
+    });
+  }
+
+  async sendInquiry(contactDto: ContactUsDto): Promise<void> {
+    // Validate email
+    if (!validator.isEmail(contactDto.email)) {
+      throw new HttpException('Invalid email address', HttpStatus.BAD_REQUEST);
+    }
```
Comment (buka4rill) [inline src/modules/contactUs/contact-us.service.ts:28]: This sort of validation is not necessary if you validate DTOs with [class validators](https://docs.nestjs.com/techniques/validation)

File: src/modules/contactUs/contact-us.service.ts [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+
+    // Sanitize data
+    const sanitizedEmail = validator.normalizeEmail(contactDto.email) || '';
+    const sanitizedMessage = validator.escape(contactDto.message);
```
Comment (buka4rill) [inline src/modules/contactUs/contact-us.service.ts:32]: Validate input DTOs instead

File: src/modules/contactUs/tests/contact-us.service.spec.ts [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+  beforeEach(async () => {
+    sendMailMock = jest.fn().mockResolvedValue(true);
+    (nodemailer.createTransport as jest.Mock).mockReturnValue({
+      sendMail: sendMailMock,
+    });
+
+    // for testing purposes
+    process.env.CONTACT_EMAIL = 'contact@example.com';
+    process.env.EMAIL_USER = 'testuser@example.com';
+    process.env.EMAIL_PASS = 'password';
```
Comment (buka4rill) [inline src/modules/contactUs/tests/contact-us.service.spec.ts:23]: What's happening here?

**Review by buka4rill — COMMENTED** [review]
Couple comments

File: src/modules/contactUs/contact-us.service.ts (hunk 7 above)
Comment (deolla) [reply src/modules/contactUs/contact-us.service.ts]: done

File: src/modules/contactUs/contact-us.service.ts (hunk 6 above)
Comment (deolla) [reply src/modules/contactUs/contact-us.service.ts]: I have made the changes

File: src/modules/contactUs/tests/contact-us.service.spec.ts (hunk 8 above)
Comment (deolla) [reply src/modules/contactUs/tests/contact-us.service.spec.ts:23]: The email configuration requires the use of environment variables. I tried using a different approach, but it didn't work.

**Comment by deolla** [comment]
done

File: src/modules/contactUs/tests/contact-us.service.spec.ts (hunk 8 above)
Comment (buka4rill) [reply src/modules/contactUs/tests/contact-us.service.spec.ts:23]: why are you defining constants like this though? I am confused

Comment (deolla) [reply src/modules/contactUs/tests/contact-us.service.spec.ts:23]: I tried using direct constants for the test, but I kept getting an error saying 'email configuration is missing'. So, I had to use environment variables instead, which resolved the issue.

**Comment by deolla** [comment]
done @buka4rill

File: src/modules/contactUs/tests/contact-us.service.spec.ts (hunk 8 above)
Comment (deolla) [reply src/modules/contactUs/tests/contact-us.service.spec.ts]: updated

File: src/modules/contactUs/contact-us.service.ts (hunk 7 above)
Comment (buka4rill) [reply src/modules/contactUs/contact-us.service.ts]: I don’t see the change

File: src/modules/contactUs/contact-us.controller.ts (hunk 1 above)
Comment (deolla) [reply src/modules/contactUs/contact-us.controller.ts:11]: I have

**Comment by Heba-WebDev** [comment]
@deolla I still see no changes. Do as the mentor told you. Your service shouldn't be concern about input validation, leave that for the DTO to deal with

**Comment by deolla** [comment]
Done @Heba-WebDev

**Review by Heba-WebDev — APPROVED** [review]

**Review by Heba-WebDev — DISMISSED** [review]

File: src/app.module.ts [hunk 9]
```diff
@@ -12,6 +12,7 @@ import { SeedingModule } from './database/seeding/seeding.module';
 import HealthController from './health.controller';
 import { AuthModule } from './modules/auth/auth.module';
 import { UserModule } from './modules/user/user.module';
+import { ContactModule } from './modules/contactUs/contact-us.module';
```
Comment (CFCIfe) [inline src/app.module.ts:15]:
```suggestion
import { ContactUsModule } from './modules/contactUs/contact-us.module';
```

File: src/app.module.ts [hunk 10]
```diff
@@ -73,6 +75,7 @@ import { TestimonialsModule } from './modules/testimonials/testimonials.module';
     SeedingModule,
     AuthModule,
     UserModule,
+    ContactModule,
```
Comment (CFCIfe) [inline src/app.module.ts:78]:
```suggestion
ContactUsModule
```

File: src/modules/contactUs/contact-us.controller.ts [hunk 11]
```diff
@@ -0,0 +1,27 @@
+import { Body, Controller, HttpStatus, Post, UsePipes, ValidationPipe } from '@nestjs/common';
+import { EmailService } from '../email/email.service';
+import { ContactUsDto } from './dto/contact-us.dto';
+import { skipAuth } from '../../helpers/skipAuth';
+import { ApiBody, ApiResponse } from '@nestjs/swagger';
+import { ContactUs } from './entities/contact-us.entity';
+
+@skipAuth()
+@Controller('/contact-us')
+export class ContactController {
```
Comment (CFCIfe) [inline src/modules/contactUs/contact-us.controller.ts:10]:
```suggestion
export class ContactUsController
```

File: src/modules/contactUs/contact-us.controller.ts [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+  constructor(private readonly contactService: EmailService) {}
+
+  @Post()
+  @UsePipes(new ValidationPipe({ whitelist: true, forbidNonWhitelisted: true }))
```
Comment (CFCIfe) [inline src/modules/contactUs/contact-us.controller.ts:14]: This can be removed. Its been applied globally in `app.module.ts`

Comment (deolla) [reply src/modules/contactUs/contact-us.controller.ts]: done @CFCIfe

File: package.json [hunk 13]
```diff
@@ -47,6 +47,7 @@
     "bcryptjs": "^2.4.3",
     "class-transformer": "^0.5.1",
     "class-validator": "^0.14.1",
+    "dotenv": "^16.4.5",
```
Comment (CFCIfe) [inline package.json:50]: Can this be taken out? We have `@nestjs/config` to handle environment variables.

File: src/modules/contactUs/contact-us.service.ts [hunk 14]
```diff
@@ -0,0 +1,39 @@
+// src/modules/contactUs/contact-us.service.ts
```
Comment (CFCIfe) [inline src/modules/contactUs/contact-us.service.ts:1]: take out unused comments please

File: src/modules/email/email.service.ts [hunk 15]
```diff
@@ -55,4 +59,13 @@ export class EmailService {
       },
     });
   }
+
+  async sendContactInquiryMail(context: { name: string; email: string; message: string; created_at: string }) {
```
Comment (CFCIfe) [inline src/modules/email/email.service.ts:63]:
```suggestion
async sendContactInquiryMail(context: emailContextDto)
```

This is a reusable dto that can also be typed at `src/modules/contactUs/contact-us.service.ts`

Comment (CFCIfe) [reply src/modules/email/email.service.ts:63]: I see you have a `ContactUsDto` already. That can be used too

**Review by CFCIfe — COMMENTED** [review]
Little comments

File: src/modules/email/email.service.ts (hunk 15 above)
Comment (deolla) [reply src/modules/email/email.service.ts]: done @CFCIfe

File: src/modules/contactUs/contact-us.service.ts (hunk 14 above)
Comment (deolla) [reply src/modules/contactUs/contact-us.service.ts]: done

File: package.json (hunk 13 above)
Comment (deolla) [reply package.json]: done @CFCIfe

**Review by Heba-WebDev — APPROVED** [review]
All good

**Review by davidolanrewaju — APPROVED** [review]
